src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/RawBinaryReader.cpp

```diff
@@ -0,0 +1,245 @@
+ * other materials provided with the distribution.
+ *
+ * Neither the name of BlueQuartz Software, the US Air Force, nor the names of its
+ * contributors may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The code contained herein was partially funded by the following contracts:
+ *    United States Air Force Prime Contract FA8650-07-D-5800
+ *    United States Air Force Prime Contract FA8650-10-D-5210
+ *    United States Prime Contract Navy N00173-07-C-2068
+ *    United States Air Force Prime Contract FA8650-15-D-5231
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#include "RawBinaryReader.hpp"
+
+#include <algorithm>
+#include <cstddef>
+
+#include "complex/Common/ComplexConstants.hpp"
+#include "complex/Common/Types.hpp"
+#include "complex/DataStructure/DataArray.hpp"
+#include "complex/DataStructure/DataStore.hpp"
+
+#if defined(_MSC_VER)
+const auto FSEEK = _fseeki64;
+#else
+const auto FSEEK = std::fseek;
+#endif
+
+namespace complex
+{
+
+namespace
+{
+#ifdef CMP_WORDS_BIGENDIAN
+constexpr int32_t k_EndianCheck = 0;
+#else
+constexpr int32_t k_EndianCheck = 1;
+#endif
+
+constexpr int32_t k_RbrNoError = 0;
+constexpr int32_t k_RbrFileNotOpen = -1000;
+constexpr int32_t k_RbrFileTooSmall = -1010;
+constexpr int32_t k_RbrFileTooBig = -1020;
+constexpr int32_t k_RbrDAError = -1040;
+constexpr int32_t k_RbrComponentError = -1050;
+constexpr int32_t k_RbrDANull = -1060;
+
+// -----------------------------------------------------------------------------
+int32_t SanityCheckFileSizeVersusAllocatedSize(size_t allocatedBytes, size_t fileSize, size_t skipHeaderBytes)
+{
+  if(fileSize - skipHeaderBytes < allocatedBytes)
+  {
+    return -1;
+  }
+  if(fileSize - skipHeaderBytes > allocatedBytes)
+  {
+    return 1;
+  }
+  // File Size and Allocated Size are equal so we are good to go
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+template <typename T>
+int32_t readBinaryFile(IDataArray* dataArrayPtr, const std::string& filename, uint64_t skipHeaderBytes, ChoicesParameter::ValueType endian)
```

Comment:
We should take the array be reference here.

```suggestion
int32 readBinaryFile(IDataArray& dataArrayPtr, const std::string& filename, uint64 skipHeaderBytes, ChoicesParameter::ValueType endian)
```